### DISCUSSION TITLE
Dispatching and multithreading

### DISCUSSION THREAD

**Discussion by etestu-kalray** [dbus-cxx/dbus-cxx#141] [Q&A] [answered]
Hi All,

After reading the doc, having a look at some of the example projects and the code of dbus-cxx, I'm still unsure what threading model is supported. Basically, what can or can't a ThreadDispatcher do?

At first, it seemed to me that you could do pretty much what you want as long as you can write a ThreadDispatcher for it.
So I've written an app with a single connection and a ThreadDispatcher based on libuv, each signal and message received is handled in libuv's threadpool and it works well.
That is, until you try to stress test it... After a while, I get a segfault in `Connection::notify_dispatcher_or_dispatch`, when calling a libsigc++ signal: `m_priv->m_needsDispatching()`.
It seems like handling signals and messages from a single connection from multiple threads at the same time is unsafe - by design or because of an oversight, I'm not sure.

A coworker had a quick look at dbus-cxx, and he thinks that there should be a thread handling messages/signals by DBus object.
That seems achievable fairly easily: one DBus object per connection, one ThreadDispatcher dispatching to one thread per connection, and as many connection as DBus objects I need.
I haven't tested it, but it shouldn't segfault because `m_priv->m_needsDispatching` wont be shared between threads.

Back to my question: is it supported for a ThreadDispatcher to dispatch to multiple threads, or am I doing something I shouldn't?

I'd be happy to open PRs for the (trivial) patch fixing the segfault when dispatching to multiple threads, and for the ThreadDispatcher if dispatching to multiple threads is a valid usage of dbus-cxx!

**Comment by rm5248** [comment, marked as answer]
> It seems like handling signals and messages from a single connection from multiple threads at the same time is unsafe - by design or because of an oversight, I'm not sure.

This should work fine, if it's not that's either due to a misuse on your side or an oversight on the dbus-cxx side.

> Back to my question: is it supported for a ThreadDispatcher to dispatch to multiple threads, or am I doing something I shouldn't?

From your question here, it sounds like you're doing something wrong.  The documentation on dispatching can be found here: https://dbus-cxx.github.io/dispatching.html

There's two ways to have methods called: they can be called from the thread that is handling the DBus connection(the dispatcher thread, which may or may not be the main thread), or they can be called from another thread in your program.  If you choose the second option, you need to have a `ThreadDispatcher` on a per-thread basis; it sounds like you have one `ThreadDispatcher` for _all_ of your threads, which is incorrect.

In my experience, most of the time you can configure dbus-cxx to call the method from the dispatcher thread, even if the dispatcher is in a separate thread.  This however does depend on how your application works, since calling across threads may result in race conditions.

**Comment by etestu-kalray** [comment]
Thank you for the quick reply!

I indeed have a single instance of `ThreadDispatcher` dispatching to a thread-pool.
If i understood correctly, a valid way of achieving that would be to instantiate a  `ThreadDispatcher` per thread of the thread-pool, and adding from all those threads the corresponding `ThreadDispatcher`s to the `Connection`?

Also, how important is it for the method to be handled in the thread expected by dbus-cxx (The `Connection` object associates a `ThreadDispatcher` with its thread ID)?

**Reply by rm5248** [reply]
> I indeed have a single instance of `ThreadDispatcher` dispatching to a thread-pool. If i understood correctly, a valid way of achieving that would be to instantiate a `ThreadDispatcher` per thread of the thread-pool, and adding from all those threads the corresponding `ThreadDispatcher`s to the `Connection`?

Yes, that should work.

> Also, how important is it for the method to be handled in the thread expected by dbus-cxx (The `Connection` object associates a `ThreadDispatcher` with its thread ID)?

This is application-dependent.  dbus-cxx will attempt to call the method from the correct thread.  For most things that I do, calling a method from the dispatcher thread is fine, so I often don't need to worry about the `ThreadDispatcher`.  The thread dispatching is just a way to help ensure that you can call methods in the correct thread - if the methods are to be called are thread-safe, it shouldn't matter.

One possible solution is instead of using the `ThreadDispatcher` would be to lock the method that is being called from dbus-cxx with a mutex to ensure that if it is called by two threads at once it won't cause an issue.

**Reply by etestu-kalray** [reply]
Again, many thanks for your quick reply!
Understandably, if a user of the lib really wants to use it in a multi-threaded way, ensuring thread-safety of its own methods is its responsibility. My interest is really what dbus-cxx can('t) do. I guess my question should have been this from the beginning:

> Assuming all the methods to be called are thread-safe, is there anything but a semantic requirement for a `ThreadDispatcher` to always cause the methods to be invoked in the same thread? Or could that thread change, should the meaning of _"the correct thread"_ change too?

From my testings, looking at the code, the documentation and this conversation, it seems to be unintended, but working just fine.

Regardless of the answer, or the phrasing of a `ThreadDispatcher` scheduling work on not-always-the-same-thread, I'll try to reproduce the issue mentioned in my first message (segfault in `Connection::notify_dispatcher_or_dispatch`) as I strongly suspect it can occur with multiple `ThreadDispatcher`s for a single connection, while under heavy load.
I'll open a PR for that next week, if this is really a dbus-cxx bug.

**Comment by etestu-kalray** [comment]
I was unable to reproduce the bug with many `ThreadDispatcher`s, each running the methods on a distinct thread. Yey!

---

For future references, if anyone feels like having a `ThreadDispatcher` running the methods in a thread-pool, this little fix was needed:
```
--- a/dbus-cxx/connection.cpp
+++ b/dbus-cxx/connection.cpp
@@ -113,6 +113,7 @@ public:
     std::mutex m_threadDispatcherLock;
     std::map<std::thread::id, std::weak_ptr<ThreadDispatcher>> m_threadDispatchers;
     std::shared_ptr<DBusDaemonProxy> m_daemonProxy;
+    std::mutex m_needsDispatchingLock;
     sigc::signal<void()> m_needsDispatching;
     std::mutex m_freeProxySignalsLock;
     std::vector<FreeSignalThreadInfo> m_freeProxySignals;
@@ -1054,6 +1055,7 @@ void Connection::notify_dispatcher_or_dispatch() {
     if( std::this_thread::get_id() == m_priv->m_dispatchingThread ) {
         dispatch();
     } else {
+        std::unique_lock<std::mutex> lock( m_priv->m_needsDispatchingLock );
         m_priv->m_needsDispatching();
     }
 }
```
--> those libsigc signals are not thread safe: you can't use them at the same time from multiple threads.
I'd suspect more patches may be needed, as dbus-cxx was not designed with that in mind.
